driver-core/src/main/com/mongodb/internal/dns/DefaultDnsResolver.java

```diff
@@ -39,6 +35,23 @@
  */
 public final class DefaultDnsResolver implements DnsResolver {
 
+    private final DnsClient dnsClient;
+
+    public DefaultDnsResolver() {
+        ServiceLoader<DnsClientProvider> loader = ServiceLoader.load(DnsClientProvider.class);
```

Comment:
This is going to be challenging to test, given the way service provider discover works.

Reply:
I think ShrinkWrap (https://arquillian.org/guides/shrinkwrap_introduction/) can be used to test it

Reply:
I tested it manually to confirm that it works as expected.  We'll have to decide whether automating it is worth the effort.

Reply:
Looks like Mockito can mock both [static methods|https://www.baeldung.com/mockito-mock-static-methods] and [final classes](https://www.baeldung.com/mockito-final). I.e., if we really want, we can write unit tests, but it also seems fine to not have a test for this code.

Reply:
So, I would recommend not following the Mockito route. I tried and got bitten (typically, impossible to use other libraries that use service loader such as assertj and so on). Unfortunately, I don't have a great solution, expect creating temporary files with the provider you want to load.